File: pixijs/game/status-system.js
// ============================================================
// STATUS SYSTEM — applyStatus(), updateStatuses(), hasStatus()
// Generic status effect system for enemies.
// Statuses are stored on enemy.statuses as { [type]: statusInstance }
// Each statusInstance: { remaining, tickTimer, tickInterval, ...options }
// ============================================================

import { Sounds } from '../core/sound.js';

// ── Status definitions ────────────────────────────────────────
// Each def: { duration, tickInterval, onApply(enemy, state, opts), onTick(enemy, state, status), onExpire(enemy, state) }

export const STATUS_DEFS = {
  freeze: {
    duration: 2.0,
    tickInterval: 0,
  },
  burn: {
    duration: 2.0,
    tickInterval: 0.2,
    onTick(enemy, state, status) {
      enemy.takeDamage(status.damagePerTick, false, { state, silent: true });

      if (enemy.hp <= 0 && !enemy.isDead) {
        enemy.die();
      }
    },
  },
};

// ── Public API ────────────────────────────────────────────────

export function applyStatus(enemy, statusType, duration, options = {}) {
  const def = STATUS_DEFS[statusType];
  if (!def) return;

  const dur = duration ?? def.duration;

  if (!enemy.statuses) enemy.statuses = {};

  const existing = enemy.statuses[statusType];
  if (existing) {
    existing.remaining = dur;
  } else {
    enemy.statuses[statusType] = {
      remaining: dur,
      tickTimer: 0,
      tickInterval: def.tickInterval,
      damagePerTick: options.damagePerTick ?? 0,
      ...options,
    };
  }

  if (def.onApply) def.onApply(enemy, options.state, options);
}

export function updateStatuses(enemy, dt, state) {
  if (!enemy.statuses) return;

  for (const type of Object.keys(enemy.statuses)) {
    const status = enemy.statuses[type];
    const def = STATUS_DEFS[type];
    if (!def) {
      delete enemy.statuses[type];
      continue;
    }

    status.remaining -= dt;
    if (status.remaining <= 0) {
      if (def.onExpire) def.onExpire(enemy, state);
      delete enemy.statuses[type];
      continue;
    }

    if (def.onTick && status.tickInterval > 0) {
      status.tickTimer += dt;
      while (status.tickTimer >= status.tickInterval) {
        status.tickTimer -= status.tickInterval;
        def.onTick(enemy, state, status);
        if (enemy.isDead) return;
      }
    }
  }
}

export function hasStatus(enemy, statusType) {
  return !!(enemy.statuses && enemy.statuses[statusType]);
}

export function clearStatuses(enemy) {
  if (enemy.statuses) enemy.statuses = {};
}

File: Scripts/Combat/StatusSystem.cs
using System.Collections.Generic;

namespace SpaceOrLife.Combat;

public enum StatusKind
{
    Freeze,
    Burn
}

public sealed class StatusEffect
{
    public float Remaining;
    public float TickTimer;
    public float TickInterval;
    public float DamagePerTick;
}

public interface IStatusTarget
{
    Dictionary<StatusKind, StatusEffect> Statuses { get; }
    bool IsDead { get; }
    void TakeDamage(float damage, bool isCrit);
}

public static class StatusSystem
{
    private const float FreezeDuration = 2f;
    private const float BurnDuration = 2f;
    private const float BurnTickInterval = 0.2f;

    public static System.Action<IStatusTarget, float>? OnBurnTick;

    public static void ApplyStatus(IStatusTarget target, StatusKind kind, float duration, float damagePerTick = 0f)
    {
        if (target.Statuses == null)
            return;

        if (target.Statuses.TryGetValue(kind, out var existing))
        {
            existing.Remaining = duration;
            if (kind == StatusKind.Burn)
                existing.DamagePerTick = damagePerTick;
        }
        else
        {
            var status = new StatusEffect
            {
                Remaining = duration,
                TickTimer = 0f,
                TickInterval = kind == StatusKind.Burn ? BurnTickInterval : 0f,
                DamagePerTick = damagePerTick,
            };
            target.Statuses[kind] = status;
        }
    }

    public static void UpdateStatuses(IStatusTarget target, float dt)
    {
        if (target.Statuses == null || target.Statuses.Count == 0)
            return;

        var keys = new List<StatusKind>(target.Statuses.Keys);
        for (int i = keys.Count - 1; i >= 0; i--)
        {
            if (target.IsDead)
                return;

            var kind = keys[i];
            var status = target.Statuses[kind];

            status.Remaining -= dt;
            if (status.Remaining <= 0f)
            {
                target.Statuses.Remove(kind);
                continue;
            }

            if (kind == StatusKind.Burn && status.TickInterval > 0f)
            {
                status.TickTimer += dt;
                while (status.TickTimer >= status.TickInterval)
                {
                    status.TickTimer -= status.TickInterval;
                    target.TakeDamage(status.DamagePerTick, false);
                    OnBurnTick?.Invoke(target, status.DamagePerTick);
                    if (target.IsDead)
                        return;
                }
            }
        }
    }

    public static bool HasStatus(IStatusTarget target, StatusKind kind)
    {
        return target.Statuses != null && target.Statuses.ContainsKey(kind);
    }

    public static void ClearStatuses(IStatusTarget target)
    {
        target.Statuses?.Clear();
    }

    public static float GetFreezeDuration() => FreezeDuration;
    public static float GetBurnDuration() => BurnDuration;
}
